traitlets/traitlets.py

```diff
@@ -680,9 +679,21 @@ def hold(*a):
                     # (only used to preserve pickleability on Python < 3.4)
                     self.__dict__.pop('_notify_trait', None)
 
-                # trigger delayed notifications
-                for v in cache.values():
-                    self._notify_trait(*v)
+    @contextlib.contextmanager
+    def hold_trait_notifications(self):
+        """Context manager for bundling trait change notifications and cross
+        validation.
+        Use this when doing multiple trait assignments (init, config), to avoid
+        race conditions in trait notifiers requesting other trait values.
+        All trait notifications will fire after all values have been assigned.
+        """
+        try:
+            with self.mute_trait_notifications() as cache:
+                yield
+        finally:
+            self._cross_validation_lock = False
```

Comment:
I'm not sure about this cross-validation logic line. I wouldn't think it should be needed since it's already restored in the finally block of mute_trait_notifications